# Tasks Assessments 2020


## Module: Machine Learning and Statistics

<br>

#### These are my solutions for the Task assessments that have been set out as apart of my ongoing studies for H. Dip in Data Analytics at GMIT.


***

## <u>Task 1: Calculate the square root of 2</u>

![Square Root image](2.jpg)
###### Image source: https://www.google.com/url?sa=i&url=https%3A%2F%2Fmagoosh.com%2Fmath%2Fpowers-and-roots-square-roots%2F&psig=AOvVaw1RujAC9Ah-YssUhQ1NOzWe&ust=1602502592305000&source=images&cd=vfe&ved=0CA0QjhxqFwoTCNiVnrm5rOwCFQAAAAAdAAAAABAD 

### The aim of this assessment is to display an algorithm that has the ability to calculate the square root of a number without the use of the built-in functions, i.e. math.sqrt.

There is documentation on the Python documentation that explains in great detail, the process and options available of how to display the decimal output based on built in modules. - Python; 9.3 decimal - Decimal fixed point and floating point arithmetic; https://docs.python.org/2/library/decimal.html


There are many different methods available to compute the square root of a number. I found alot of indepth information on these different options and methods, documented on on this link: https://en.wikipedia.org/wiki/Methods_of_computing_square_roots
<br>

<i>"The method employed depends on what the result is to be used for (i.e. how accurate it has to be), how much effort one is willing to put into the procedure, and what tools are at hand."</i>
<br>

The method that you choose also depends on the way in which the calculations are going to be completed. Some methods can be completed mentally or via paper-and-pencil whereas others require indepth calculations by computation.
<br>

#### How float numbers are stored in memory on a computer?

Having researched the aspect of how a floating point number is stored on a computer, the area is an extremely vast and complex one. 
<br>

From my understanding of how the memory storage takes place, the bits are broken up into 3 areas which are used for the storage of: the sign, the mantissa and the expotent.
<br>
<i>"The mantissa holds the main digits and the exponents defines where the decimal point should be placed. The same technique can be used for binary numbers." </i> https://en.wikibooks.org/wiki/A-level_Computing/AQA/Paper_2/Fundamentals_of_data_representation/Floating_point_numbers
<br>

The standard for Floating-Point arithmetic is set out by the IEEE (Institute of Electrical and Electronics Engineers).

<i>"The IEEE Standard for Floating-Point Arithmetic (IEEE 754) is a technical standard for floating-point arithmetic established in 1985"</i> https://en.wikipedia.org/wiki/IEEE_754
<br>
Over time, there have been edits make to the standard with the latest one published and released in 2019.
<br>

<i>"The current version, IEEE 754-2019, was published in July 2019.[1] It is a minor revision of the previous version, incorporating mainly clarifications, defect fixes and new recommended operations."</i> https://en.wikipedia.org/wiki/IEEE_754
<br>

In my research for this task, I have discovered that this is not the first time that this query has come about and found various threads on coding websites requesting assistance with this same task. Having read the following links, I can see that some codes are extremely indepth and of a high standard and some are in various other languages (e.g. C).
<ol>
    <li> Stackoverflow - Is there a way to create more decimal points on Python without importing a library/module?: https://stackoverflow.com/questions/64278117/is-there-a-way-to-create-more-decimal-points-on-python-without-importing-a-libra </li>\
    <li> Stackoverflor - How to compute the digits of an irrational number one by one?:
 https://stackoverflow.com/questions/60318091/how-to-compute-the-digits-of-an-irrational-number-one-by-one </li>
    <li> Stack Exchange - Can anyone explain representation of float in memory?: 
https://softwareengineering.stackexchange.com/questions/215065/can-anyone-explain-representation-of-float-in-memory#:~:text=Floating%2Dpoint%20numbers%20are%20encoded,along%20with%20a%20sign%20bit).&text=Like%20signed%20integer%20types%2C%20the,are%20used%20for%20the%20exponent </li>
</ol>

I also researched this issue using various sites to gain a better understanding of the task:
<ol>
    <li> Keil.com; Floating-point Numbers; https://www.keil.com/support/man/docs/c51/c51_ap_floatingpt.htm#:~:text=Scalars%20of%20type%20float%20are,follows%20the%20IEEE%2D754%20standard.&text=The%20mantissa%20represents%20the%20actual,of%20the%20floating%2Dpoint%20number.&text=The%20actual%20value%20of%20the,of%20%E2%80%93127%20to%20%2B128 </li>
    <li> Log2Base2.com; How float or double values are stored in memory? https://www.log2base2.com/storage/how-float-values-are-stored-in-memory.html </li>
    <li> Floating Point Numbers; https://users.cs.fiu.edu/~downeyt/cop2400/float.htm </li>
</ol>


In [ ]:
# An example of how to compute the required task using a built in-function
## https://stackoverflow.com/questions/4733173/how-can-i-show-an-irrational-number-to-100-decimal-places-in-python

from decimal import *
getcontext().prec = 100
Decimal(2).sqrt()

In [ ]:
## Another method of setting the decimal output using a built-in function 
## https://stackoverflow.com/questions/4733173/how-can-i-show-an-irrational-number-to-100-decimal-places-in-python

import gmpy2
ctx = gmpy2.get_context()
ctx.precision = 100
print(gmpy2.sqrt(2))

## The output however does not output 100 decimal places as requested in the function
## whn I chose to run this function again with 200 places it also displays less than requested. 
## This could be linked to - https://docs.python.org/3.4/tutorial/floatingpoint.html

In [ ]:
# https://www.programiz.com/python-programming/examples/square-root
# Python Program to calculate the square root

# Note: change this value for a different result
num = 2

# To take the input from the user
#num = float(input('Enter a number: '))

num_sqrt = num ** 0.5
print('The square root of %0.3f is %0.3f'%(num ,num_sqrt))

In [ ]:
# https://www.tutorialspoint.com/How-to-perform-square-root-without-using-math-module-in-Python
# Example of using an expotentiation operator to gain the initial request of this task

2**(1/2)

In [15]:
# https://web.microsoftstream.com/video/4199579c-c3ed-48b7-9c5c-2911cefff50b
# https://www.w3schools.com/python/python_functions.asp
# https://www.programiz.com/python-programming/examples/square-root
# https://www.tutorialspoint.com/How-to-perform-square-root-without-using-math-module-in-Python


def sqrt2(n):
    return n**0.5

for x in range (2, 101):
    ##print(sqrt2(round(x, 100))
    ## I was attempting to produce an output with rounding the result to 100 decimal places as per info on site:
    ## https://stackoverflow.com/questions/44422918/square-root-to-2-decimal-places
    print(format (sqrt2(x,'.2f'))      
    
    ## print("{:.100f}".format(number))
    ## Another alternative I tried but unable to make the code work
          


SyntaxError: unexpected EOF while parsing (<ipython-input-15-03d7337e61f9>, line 15)

In [ ]:
# https://www.sololearn.com/Discuss/2228331/how-can-we-calculate-square-root-without-using-any-built-in-function-in-python
# Example of a simple function to gain the initial request of this task (without the set decimal places)

# square root
a=int(input("enter no."))
b=a**0.5
print(b)

I have made various attempts to create and set decimal places manually through the code I create rather than relying on an import function or module.
I came across options that are not relevant in this instance:
<ul>
    <li> https://kodify.net/python/math/round-decimals/ </li>
    <li> https://www.kite.com/python/answers/how-to-print-a-float-with-two-decimal-places-in-python#:~:text=Use%20str.,string%20to%20print%20the%20float. </li>
</ul>

In [17]:
## Perfect example of the code required for this task 
## https://stackoverflow.com/questions/64278117/is-there-a-way-to-create-more-decimal-points-on-python-without-importing-a-libra

x = 2 * 10 ** 200

r = x

def test_diffs(x, r):
    d0 = abs(x - r**2)
    dm = abs(x - (r-1)**2)
    dp = abs(x - (r+1)**2)
    minimised = d0 <= dm and d0 <= dp
    below_min = dp < dm
    return minimised, below_min

while True:
    oldr = r
    r = (r + x // r) // 2

    minimised, below_min = test_diffs(x, r)
    if minimised:
        break

    if r == oldr:
        if below_min:
            r += 1
        else:
            r -= 1
        minimised, _ = test_diffs(x, r)
        if minimised:
            break

print(f'{r // 10**100}.{r % 10**100:0100d}')

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


#### <u>Conclusion:</u>

Having spent a considerable time trying to fathom how to create this code, I have learned to appreciate the ease and convenience that comes with the built-in codes and libraries that make coding such as this quite straightforward. 
There appears to be alot of attempts in the past to achieve the same task as myself, based on the results found online throughout my research. I look forward to building on my knowledge and info to date to overcome this challenge in the future while also using this as a benchmark to attempt and complete other coding without the use of built-in libraries. 

#### <u>References:</u>

**************************
**************************


## <u>Task 2: Chi-squared test for independence</u>

<br>

#### <u>Background:</u>

The chi-squared test was developed during the 19th Century by Karl Pearson. Pearson identified skewness within biological observations and <i>"developed the Pearson distribution proposing a method to model the observation and perform a test of goodness of fit to determine how well the model really fits to the observations."</i> https://en.wikipedia.org/wiki/Chi-squared_test

The aim of the chi-squared test is to determine if there is a relation to the variables within the categories of the contingency table (also known as a crosstab).
<br>


![Chi-squared test table](chi-table.PNG)

In [19]:
import numpy as np
import pandas as pd
import scipy.stats as stats
from scipy.stats import chi2_contingency

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# load the dataset
# df = get_data('filename').data
# df.head()
# https://reneshbedre.github.io/blog/chisq.html
# https://www.kaggle.com/hamelg/python-for-data-25-chi-squared-tests

df = pd.DataFrame({'Neighbourhood' : ['A', 'B', 'C', 'D'] * 4, 'Collar' : ['White', 'Blue', 'NoCollar'] * 4 })

df.head()

## Attempt to run using info from https://predictivehacks.com/how-to-run-chi-square-test-in-python/ failed 
## as per various errors as the dataframe I attempted to create had 4 pieces of info in 1 array and 3 in the other. 


ValueError: arrays must all be same length

In [5]:
import numpy as np
import pandas as pd
import scipy.stats as stats
from scipy.stats import chi2_contingency

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

data = pd.read_csv("chi.csv")

print (data.head(4))


       Category    A    B    C    D  TOTAL
0  White collar   90   60  104   95    349
1   Blue collar   30   50   51   20    151
2     No Collar   30   40   45   35    150
3         Total  150  150  200  150    650


In [6]:
data.describe()

A           B           C           D       TOTAL
count    4.000000    4.000000    4.000000    4.000000    4.000000
mean    75.000000   75.000000  100.000000   75.000000  325.000000
std     57.445626   50.662281   71.744919   59.581876  236.009887
min     30.000000   40.000000   45.000000   20.000000  150.000000
25%     30.000000   47.500000   49.500000   31.250000  150.750000
50%     60.000000   55.000000   77.500000   65.000000  250.000000
75%    105.000000   82.500000  128.000000  108.750000  424.250000
max    150.000000  150.000000  200.000000  150.000000  650.000000

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Category  4 non-null      object
 1   A         4 non-null      int64 
 2   B         4 non-null      int64 
 3   C         4 non-null      int64 
 4   D         4 non-null      int64 
 5   TOTAL     4 non-null      int64 
dtypes: int64(5), object(1)
memory usage: 320.0+ bytes


In [8]:
data.dtypes

Category    object
A            int64
B            int64
C            int64
D            int64
TOTAL        int64
dtype: object

In [12]:
pd.read_csv("chi.csv")

from scipy.stats import chi2_contingency
obs = np.array("chi.csv")
chi2_contingency(obs)

## https://www.kaggle.com/jfeng1023/beginner-data-challenge-chi-square-test


TypeError: '<' not supported between instances of 'numpy.ndarray' and 'int'

In [18]:
# https://reneshbedre.github.io/blog/chisq.html
# https://stackoverflow.com/questions/11725115/p-value-from-chi-sq-test-statistic-in-python

from scipy.stats.distributions import chi2
chi2.sf(24.6, 0.5)

1.8086220862906192e-07

##### Automated calculator for chi square
Calculate the chi square test output of 24.6
https://www.socscistatistics.com/tests/chisquare2/default2.aspx
Found a website allowing to manually enter the info for the site to calculate it
The calculation was correct as per the task information

![Chi-square calculator](sample_p.PNG)



####  <u>Conclusion:</u>

I found this task tough in relation to locating information online for assisting with trying to use the provided table of data for the task.
Any info for feeding in data manually required the same number of entries in each array whereas we had 4 population areas and 3 different collar options. 
Also when trying to create a contingency table, the data is usually fed in manually based on the info I found during my research, versus reading in a csv file.
I chose to read in the data via a CSV file which I created in Excel which allowed me to produce some output of data. 
<br>
I believe that there is so much more to this task that could be explored in more depth if time contraints were not an issue. I believe that even with a small amount of data in the CSV file, the chi square test can produce quite alot of data and output that can be easily discussed and contrasted to compare the likelyhood of expectancies from the test. 
<br>
The test itself is to determine whether there is <i> "a statistically significant difference between the expected frequencies and the observed frequencies in one or more categories of a contingency table."</i> https://en.wikipedia.org/wiki/Chi-squared_test
    
#### <u>References:</u>
    
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html



***
***

## <u>Task 3: Difference between the Excel functions: STDEV.S and STDEV.P</u>

<br>

**<i>"The standard deviation is a measure that indicates how much the values of the set of data deviate (spread out) from the mean. To put it differently, the standard deviation shows whether your data is close to the mean or fluctuates a lot."**</i> [1] 

<ul>
    <li> Based on the above explaination, the lower that the standard deviation is or the closer it is to 0, the more reliable the mean is. </li>
    <li> If the standard deviation equals to 0 exactly, this result would mean that every element in the data set is the same as its mean. </li>
    <li> The higher the standard deviation, the more fluctuation there is within the data, thus resulting in the mean being less accurate. </li>
</ul>

## What is the main difference between STDEV.S and STDEV.P?

The STDEV.S function is used to calculate the standard deviation of a sample where only a subset of the data is included. <br>
The STDEV.P function is used to calculate the standard deviation of a population which includes all elements of a data set.

![STDEV formulas](STDEV.PNG) [1]


### STDEV.S Function

The STDEV.S function was introduced in Excel 2010 as an improvement on its predecessor, the STDEV function. <br>
STDEV (now an outdated function) was introduced in Excel 2003 and is the oldest Excel function used to estimate the standard deviation of a sample. 

The method to obtain the deviation of the sample standard deviation differs from the population standard deviation because you minus 1 from the total squared differences. 
The explaination of why this is the case as per [1] is, <strong>"you need to correct the bias in the estimation of a sample mean instead of the true population mean. This is also referred to as the Bessel's correction."</strong>


### STDEV.P Function

Similar to STDEV.S, the STDEV.P function was introduced in Excel 2010 as the modern version of the STDEVP function. It is reported to provide improved accuracy and it is available in Excel 2010 and later versions. <br>
The STDEV.P function <strong>"calculates standard deviation using the "n" method, ignoring logical values and text."</strong> [2]


![STDEV differences](stdev_table.PNG)


#### A simplified guide of how to calculate the standard deviation:
<ol>
    <li>Calculate the mean of all values in the data set</li>
    <li>Subtract the mean from each number and square the result</li>
    <li>Add up squared differences</li>
    <li>Divide the total squared differences by the count of values</li>
    <li>Calculate the square root of the number obtained in previous step</li> 
</ol>


In [29]:
# https://datatofish.com/read_excel/


import pandas as pd

df = pd.read_excel (r'C:\Users\valer\Desktop\Tasks.repo\Machine_Learning_Tasks\STDEV.xlsx')
print (df)

# I created a sample file within Excel to use for this task
# To calculate the Population SD I used the following formula in Excel: =STDEV.P(B2:B13)
# To calculate the Sample SD I used the following formula in Excel: =STDEV.S(B2:B7)

      NAME  MOODLE RESULT                            .1
0     Alex             71        Population SD  13.1053
1     John             82            Sample SD  9.13053
2     Mary             70                              
3      Dan             62                              
4    Ralph             78                              
5      Bob             58                              
6   Darren             49                              
7    Sarah             76                              
8    Ciara             82                              
9     Mark             45                              
10    Sean             51                              
11  Sandra             50                              


In [15]:
# https://www.codegrepper.com/code-examples/python/numpy+STDEV.S

# The following code shows the calculation of the standard deviation of a population
# But it does not calculate accurately for a sample size
# I chose to use the same figures within the data I used in the Excel file so as to allow ease of comparison of results

import numpy as np
data = [71,82,70,62,78,58]
data_std = np.std(data)
print(data_std)

# Another 2 code examples from the same site of calculating the standard deviation using numpy which outputs the same result as the above code.

import numpy as np
values=[71,82,70,62,78,58]
print(np.std(values))

values=[71,82,70,62,78,58,np.nan]
print(np.nanstd(values))

8.334999833366657
8.334999833366657
8.334999833366657


In [31]:
# The same result is seen using this code as the above. Both seem to calculate the standard deviation of a population with no
# reference to considering the standard deviation of a sample.
# https://stackoverflow.com/questions/34050491/standard-deviation-in-numpy

np.std([71,82,70,62,78,58])

8.334999833366657

In [33]:
# When I use the same code for the full dataset it calculates the standard deviation of the population accurately.

np.std([71,82,70,62,78,58,49,76,82,45,51,50])

13.105342422081156

In [34]:
import numpy as np
values=[71,82,70,62,78,58,49,76,82,45,51,50]
print(np.std(values))

13.105342422081156


In [12]:
# https://honingds.com/blog/python-standard-deviation/
# https://stackoverflow.com/questions/34050491/standard-deviation-in-numpy

# I calculated the standard deviation of the sample data using the ddof parameter in the code and this produces the same output
# as the Excel formula STDEV.S

np.std([71,82,70,62,78,58], ddof=1)

9.130534851073438

In [19]:
x = [71,82,70,62,78,58,49,76,82,45,51,50]
np.sqrt(np.sum((x - np.mean(x))**2)/len(x))

13.105342422081156

In [36]:
# When I use the numpy formula for the sample data set, the output produced is different to Excel result

x = [71,82,70,62,78,58]
np.sqrt(np.sum((x - np.mean(x))**2)/len(x)-1)

8.274794391537606

#### <u>Conclusion:</u>

Based on my research for this task along with the results observed from the above calculations, I am of the opinion that the Excel formula "STDEV.S" is a more accurate and reliable method to calculate the standard deviation of a sample from a dataset. <br>
The formula used within the Excel application is an improvement on a previously

When I calculated the sample standard deviation using np.std with the ddof parameter, the result was accurate and matched the Excel output of STDEV.S. However the parameter needed to be inserted in order for the formula to make the assumption to allow for the reduction of the mean. <br>
Seperately to the computer science side of this task, I came across an article that stated a relatively easy explaination of how to estimate standard deviation. <br> <br>
<strong><i>"The range rule tells us that the standard deviation of a sample is approximately equal to one-fourth of the range of the data."</strong></i> [8] <br> <br>
I have not yet explored this aspect of the research further but it is certainly an interesting opinion.

#### <u>References:</u>

<ol>
    <li>https://www.ablebits.com/office-addins-blog/2017/05/31/calculate-standard-deviation-excel/</li>
    <li>https://exceljet.net/excel-functions/excel-stdev.p-function#:~:text=The%20STDEV.,deviation%20for%20an%20entire%20population.</li>
    <li>https://corporatefinanceinstitute.com/resources/excel/functions/stdev-function/#:~:text=The%20STDEV%20Function%20is%20categorized,deviation%20based%20on%20a%20sample.</li>
    <li>https://datatofish.com/read_excel/</li>
    <li>https://www.codegrepper.com/code-examples/python/numpy+STDEV.S</li>
    <li>https://stackoverflow.com/questions/34050491/standard-deviation-in-numpy</li>
    <li>https://honingds.com/blog/python-standard-deviation/</li>
    <li>https://www.thoughtco.com/range-rule-for-standard-deviation-3126231#:~:text=The%20range%20rule%20tells%20us,estimate%20of%20the%20standard%20deviation.</li>
</ol>

## <u>Task 4: Use scikit-learn to apply k-means clustering to Fisher’s famous Iris data set</u>

![Iris_image](iris.PNG)

#### <u>Background:</u>

The dataset used for this task is one created by R.A Fishera that relates to 3 different species of Iris flowers and four measurements relating to the length and width of the sepals and petals. <br>
The dataset is made up of 150 observations, consisting of 3 varieties of the Iris flower, with the measurements making up the:
<ul>
    <li>Sepal Length</li>
    <li>Sepal Width</li>
    <li>Petal Length</li>
    <li>Petal Width</li>
</ul>


Further information relating to the existence of the dataset can be reviewed here: https://en.wikipedia.org/wiki/Iris_flower_data_set


##### What is Clustering?

Clustering refers to the grouping together of a collection of data points based on certain similarities which allows to identify any potential underlying patterns.
K-means is used to look for any clusters that may exist within a dataset. 


<i>"K-means clustering is one of the simplest and popular unsupervised machine learning algorithms.
Typically, unsupervised algorithms make inferences from datasets using only input vectors without referring to known, or labelled, outcomes."</i> https://towardsdatascience.com/understanding-k-means-clustering-in-machine-learning-6a6e67336aa1

##### How the process of K-Means Clustering works?

![KMeans_explaination](KMeans-explaination.PNG)

As per the included image, the steps relating to the process of this method are as follows:
<ul>
    <ol> K means are randomly generated within the data domain</ol>
    <ol> K Clusters are created by linking / associating each observation with the nearest mean.</ol>
    <ol> The <i>centroid</i> of each k cluster then becomes the new mean.</ol>
    <ol> Steps 2 and 3 are repeated until the point of <i>convergence</i> is achieved.</ol>
</ul>

#### <u>Research:</u>

In [3]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

data = pd.read_csv("iris.csv")
print(data.head(6))

   sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa
5           5.4          3.9           1.7          0.4  Setosa


In [4]:
X = data.drop('variety', 1)
X.head(6)

sepal.length  sepal.width  petal.length  petal.width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2
5           5.4          3.9           1.7          0.4

In [6]:
from sklearn.preprocessing import MinMaxScaler

mms = MinMaxScaler()
mms.fit(data)
data_transformed = mms.transform(data)

# Attempted to use this code as per https://blog.cambridgespark.com/how-to-determine-the-optimal-number-of-clusters-for-k-means-clustering-14f27070048f
# But could not remove the variety line from the data, will revisit this at a later dat to try resolve

ValueError: could not convert string to float: 'Setosa'

#### <u>Conclusion:</u>


#### <u>References:</u>
<ul>
    <li> https://towardsdatascience.com/understanding-k-means-clustering-in-machine-learning-6a6e67336aa1 </li>
    <li>https://scikit-learn.org/stable/auto_examples/datasets/plot_iris_dataset.html</li>
    <li>https://archive.ics.uci.edu/ml/datasets/iris</li>
    <li>https://gist.github.com/curran/a08a1080b88344b0c8a7#file-iris-csv</li>
    <li>https://github.com/ybenzaki/kmeans-iris-dataset-python-scikit-learn/blob/master/kmeans%20clustering.ipynb</li>
    <li>https://github.com/ybenzaki/kmeans-iris-dataset-python-scikit-learn/blob/master/kmeans_clustering.py</li>
    <li>https://github.com/bhattbhavesh91/k_means_iris_dataset</li>
    <li>https://gist.github.com/naufraghi/457394</li>
    <li>https://github.com/amanRaina/K-means-clustering</li>
    <li>https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data</li>
    <li>https://en.wikipedia.org/wiki/Iris_flower_data_set</li>
</ul>
